sql/database.go

```diff
@@ -202,64 +225,79 @@ func InMemory(opts ...Opt) *sqliteDatabase {
 	var flags sqlite.OpenFlags
 	if !config.forceFresh {
 		flags = sqlite.SQLITE_OPEN_READWRITE |
-			sqlite.SQLITE_OPEN_WAL |
 			sqlite.SQLITE_OPEN_URI |
 			sqlite.SQLITE_OPEN_NOMUTEX
+		if !config.temp {
+			// Note that SQLITE_OPEN_WAL is not handled by SQLITE api itself,
+			// but rather by the crawshaw library which executes
+			// PRAGMA journal_mode=WAL in this case.
+			// We don't want it for temporary databases as they're not
+			// using any journal
+			flags |= sqlite.SQLITE_OPEN_WAL
+		}
 	}
 	freshDB := config.forceFresh
-	pool, err := sqlitex.Open(uri, flags, config.connections)
+	pool, err := sqlitex.Open(config.uri, flags, config.connections)
 	if err != nil {
 		if config.forceFresh || sqlite.ErrCode(err) != sqlite.SQLITE_CANTOPEN {
-			return nil, fmt.Errorf("open db %s: %w", uri, err)
+			return nil, fmt.Errorf("open db %s: %w", config.uri, err)
 		}
 		flags |= sqlite.SQLITE_OPEN_CREATE
 		freshDB = true
-		pool, err = sqlitex.Open(uri, flags, config.connections)
+		pool, err = sqlitex.Open(config.uri, flags, config.connections)
 		if err != nil {
-			return nil, fmt.Errorf("create db %s: %w", uri, err)
+			return nil, fmt.Errorf("create db %s: %w", config.uri, err)
 		}
 	}
 	db := &sqliteDatabase{pool: pool}
 	if config.enableLatency {
 		db.latency = newQueryLatency()
 	}
+
+	if config.temp {
+		// Temporary database is used for migration and is deleted if migrations
+		// fail, so we make it faster by disabling journaling and synchronous
+		// writes.
+		if _, err := db.Exec("PRAGMA journal_mode=OFF", nil, nil); err != nil {
+			return nil, errors.Join(
+				fmt.Errorf("PRAGMA journal_mode=OFF: %w", err),
+				db.Close())
+		}
+		if _, err := db.Exec("PRAGMA synchronous=OFF", nil, nil); err != nil {
+			return nil, errors.Join(
+				fmt.Errorf("PRAGMA journal_mode=OFF: %w", err),
+				db.Close())
+		}
+	}
+
 	if freshDB && !config.forceMigrations {
 		if err := config.schema.Apply(db); err != nil {
 			return nil, errors.Join(
 				fmt.Errorf("error running schema script: %w", err),
 				db.Close())
 		}
-	} else {
-		before, after, err := config.schema.CheckDBVersion(logger, db)
-		switch {
-		case err != nil:
-			return nil, errors.Join(err, db.Close())
-		case before != after && config.enableMigrations:
-			logger.Info("running migrations",
-				zap.Int("current version", before),
-				zap.Int("target version", after),
-			)
-			if err := config.schema.Migrate(
-				logger, db, before, config.vacuumState,
-			); err != nil {
-				return nil, errors.Join(err, db.Close())
-			}
-		case before != after:
-			logger.Error("database version is too old",
-				zap.Int("current version", before),
-				zap.Int("target version", after),
-			)
-			return nil, errors.Join(
-				fmt.Errorf("%w: %d < %d", ErrOldSchema, before, after),
-				db.Close())
+	} else if db, err = ensureDBSchemaUpToDate(logger, db, config); err != nil {
+		if db != nil {
+			err = errors.Join(err, db.Close())
 		}
+		return nil, err
```

Comment:
Why does `ensureDBSchemaUpToDate` take `db` and return `db`? Why does the caller need to close something that is returned with an error?

Reply:
`ensureDBSchemaUpToDate` may replace the original database and open the new one.
In other cases, when vacuuming is not needed, it doesn't open or close any databases, and if the migrations fail, we need to close the db. I'll add a clarifying comment for that.

Reply:
Added a comment

Reply:
Also, removed `errors.Join`